core/server/proxy/src/main/java/alluxio/proxy/s3/MultipartUploadCleaner.java

```diff
@@ -0,0 +1,288 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.proxy.s3;
+
+import alluxio.AlluxioURI;
+import alluxio.client.file.FileSystem;
+import alluxio.client.file.URIStatus;
+import alluxio.conf.PropertyKey;
+import alluxio.conf.ServerConfiguration;
+import alluxio.exception.AlluxioException;
+import alluxio.exception.FileDoesNotExistException;
+import alluxio.exception.InvalidPathException;
+import alluxio.grpc.DeletePOptions;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A lazy method (not scan the whole fileSystem to find tmp directory) to
+ * support automatic abortion s3 multipartUpload after a timeout.
+ */
+public class MultipartUploadCleaner {
+  private static final Logger LOG = LoggerFactory.getLogger(MultipartUploadCleaner.class);
+
+  private final long mTimeout;
+  private final int mRetry;
+  private final long mRetryDelay;
+  private final FileSystem mFs;
+  private final ScheduledThreadPoolExecutor mExecutor;
+  private final ConcurrentHashMap<AbortTask, ScheduledFuture<?>> mTasks =
+      new ConcurrentHashMap<>();
+
+  /**
+   * Creates a new instance of {@link MultipartUploadCleaner}.
+   *
+   * @param fs the {@link FileSystem} instance
+   */
+  public MultipartUploadCleaner(FileSystem fs) {
+    mTimeout =
+        ServerConfiguration.getMs(PropertyKey.PROXY_S3_MULTIPART_UPLOAD_TIMEOUT);
+    mRetry =
+        ServerConfiguration.getInt(PropertyKey.PROXY_S3_MULTIPART_UPLOAD_CLEANER_RETRY_COUNT);
+    mRetryDelay =
+        ServerConfiguration.getMs(PropertyKey.PROXY_S3_MULTIPART_UPLOAD_CLEANER_RETRY_DELAY);
+    mExecutor = new ScheduledThreadPoolExecutor(
+        ServerConfiguration.getInt(PropertyKey.PROXY_S3_MULTIPART_UPLOAD_CLEANER_POOL_SIZE));
+
+    mFs = fs;
+  }
+
+  /**
+   * Schedule a task to clean multipart upload.
+   *
+   * @param bucket bucket name
+   * @param object object name
+   * @return true if add a abort task
+   */
+  public boolean apply(String bucket, String object)
+      throws IOException, AlluxioException {
+    // Use schedule pool do everyThing.
+    long delay = tryAbortMultipartUpload(bucket, object, null);
+    if (delay > 0) {
+      long uploadId = getMultipartUploadId(bucket, object);
```

Comment:
This can result in NPE if `getMultipartUploadId` returns null. We should either prevent that method from returning null or handle that here